board/usb.h

```diff
@@ -84,37 +84,77 @@ int  usb_cb_ep1_in(uint8_t *usbdata, int len, int hardwired);
 void usb_cb_ep2_out(uint8_t *usbdata, int len, int hardwired);
 void usb_cb_ep3_out(uint8_t *usbdata, int len, int hardwired);
 
+// descriptor types
+// same as setupdat.h
+#define DSCR_DEVICE_TYPE 1
+#define DSCR_CONFIG_TYPE 2
+#define DSCR_STRING_TYPE 3
+#define DSCR_INTERFACE_TYPE 4
+#define DSCR_ENDPOINT_TYPE 5
+#define DSCR_DEVQUAL_TYPE 6
+
+// for the repeating interfaces
+#define DSCR_INTERFACE_LEN 9
+#define DSCR_ENDPOINT_LEN 7
+#define DSCR_CONFIG_LEN 9
+#define DSCR_DEVICE_LEN 18
+
+// endpoint types
+#define ENDPOINT_TYPE_CONTROL 0
+#define ENDPOINT_TYPE_ISO 1
+#define ENDPOINT_TYPE_BULK 2
+#define ENDPOINT_TYPE_INT 3
+
+//Convert machine byte order to USB byte order
+#define TOUSBORDER(num)\
+  (num&0xFF), ((num>>8)&0xFF)
+
 uint8_t device_desc[] = {
-  0x12,0x01,0x00,0x01,
-  0xFF,0xFF,0xFF,0x40,
-  (USB_VID>>0)&0xFF,(USB_VID>>8)&0xFF,
-  (USB_PID>>0)&0xFF,(USB_PID>>8)&0xFF,
+  DSCR_DEVICE_LEN, DSCR_DEVICE_TYPE, 0x00, 0x01, //Length, Type, bcdUSB
+  0xFF, 0xFF, 0xFF, 0x40, // Class, Subclass, Protocol, Max Packet Size
+  TOUSBORDER(USB_VID), // idVendor
+  TOUSBORDER(USB_PID), // idProduct
 #ifdef STM32F4
-  0x00,0x23,
+  0x00, 0x23, // bcdDevice
 #else
-  0x00,0x22,
+  0x00, 0x22, // bcdDevice
 #endif
-  0x01,0x02,
-  0x03,0x01};
+  0x01, 0x02, // Manufacturer, Product
+  0x03, 0x01 // Serial Number, Num Configurations
+};
+
+#define ENDPOINT_RCV 0x80
+#define ENDPOINT_SND 0x00
 
 uint8_t configuration_desc[] = {
-  0x09, 0x02, 0x27, 0x00,
-  0x01, 0x01, 0x00, 0xc0,
-  0x32,
+  DSCR_CONFIG_LEN, DSCR_CONFIG_TYPE, // Length, Type,
+  0x27, 0x00, // Total Len (uint16)
+  0x01, 0x01, 0x00, // Num Interface, Config Value, Configuration
+  0xc0, 0x32, // Attributes, Max Power
   // interface 0
-  0x09, 0x04, 0x00, 0x00,
-  0x03, 0xff, 0xFF, 0xFF,
-  0x00,
+  DSCR_INTERFACE_LEN, DSCR_INTERFACE_TYPE, // Length, Type
+  0x00, 0x00, 0x03, // Index, Alt Index idx, Endpoint count
+  0xff, 0xFF, 0xFF, // Class, Subclass, Protocol
+  0x00, // Interface
   // endpoint 1, read CAN
-  0x07, 0x05, 0x81, 0x02, 0x40, 0x00, 0x00,
+  DSCR_ENDPOINT_LEN, DSCR_ENDPOINT_TYPE, // Length, Type
+  ENDPOINT_RCV | 1, ENDPOINT_TYPE_INT, // Endpoint Num/Direction, Type
+  TOUSBORDER(0x0040), // Max Packet (0x0040)
+  0x0A, // Polling Interval (10 frames)
```

Comment:
This can be anything as long as the kernel is able to handle it. I will set it to 5 and check that it works.